library/kani/src/slice.rs

```diff
@@ -38,50 +39,105 @@ fn any_range<const LENGTH: usize>() -> (usize, usize) {
+/// struct.
 ///
 /// # Example:
 ///
 /// ```rust
-/// let slice: kani::slice::NonDetSlice<u8, 5> = kani::slice::any_slice();
+/// let slice: kani::slice::AnySlice<u8, 5> = kani::slice::any_slice();
 /// foo(&slice); // where foo is a function that takes a slice and verifies a property about it
 /// ```
-pub struct NonDetSlice<T, const MAX_SLICE_LENGTH: usize> {
-    arr: [T; MAX_SLICE_LENGTH],
+pub struct AnySlice<T, const MAX_SLICE_LENGTH: usize> {
+    layout: Layout,
+    ptr: *mut T,
     slice_len: usize,
 }
 
-impl<T, const MAX_SLICE_LENGTH: usize> NonDetSlice<T, MAX_SLICE_LENGTH> {
-    fn new() -> Self {
-        let arr: [T; MAX_SLICE_LENGTH] = unsafe { any_raw() };
+impl<T, const MAX_SLICE_LENGTH: usize> AnySlice<T, MAX_SLICE_LENGTH> {
+    fn new() -> Self
+    where
+        T: Arbitrary,
+    {
         let slice_len: usize = any();
         assume(slice_len <= MAX_SLICE_LENGTH);
-        Self { arr, slice_len }
+        let layout = Layout::array::<T>(slice_len).unwrap();
+        let ptr = unsafe { alloc(layout) };
+        unsafe {
+            let mut i = 0;
+            // Note: even though the guard `i < MAX_SLICE_LENGTH` is redundant
+            // since the assumption above guarantees that `slice_len` <=
+            // `MAX_SLICE_LENGTH`, without it, CBMC fails to infer the required
+            // unwind value, and requires specifying one, which is inconvenient.
+            // CBMC also fails to infer the unwinding if the loop is simply
+            // written as:
+            //     for i in 0..slice_len {
+            //         *(ptr as *mut T).add(i) = any();
+            //     }
+            while i < slice_len && i < MAX_SLICE_LENGTH {
+                *(ptr as *mut T).add(i) = any();
+                i += 1;
+            }
+        }
+        Self { layout, ptr: ptr as *mut T, slice_len }
+    }
+
+    fn new_raw() -> Self {
+        let slice_len: usize = any();
+        assume(slice_len <= MAX_SLICE_LENGTH);
+        let layout = Layout::array::<T>(slice_len).unwrap();
+        let ptr = unsafe { alloc(layout) };
+        unsafe {
+            let mut i = 0;
+            // See note on `MAX_SLICE_LENGTH` in `new` method above
+            while i < slice_len && i < MAX_SLICE_LENGTH {
+                *(ptr as *mut T).add(i) = any_raw();
+                i += 1;
+            }
+        }
+        Self { layout, ptr: ptr as *mut T, slice_len }
     }
 
     pub fn get_slice(&self) -> &[T] {
-        &self.arr[..self.slice_len]
+        unsafe { std::slice::from_raw_parts(self.ptr, self.slice_len) }
     }
 
     pub fn get_slice_mut(&mut self) -> &mut [T] {
-        &mut self.arr[..self.slice_len]
+        unsafe { std::slice::from_raw_parts_mut(self.ptr, self.slice_len) }
     }
 }
 
-impl<T, const MAX_SLICE_LENGTH: usize> Deref for NonDetSlice<T, MAX_SLICE_LENGTH> {
+impl<T, const MAX_SLICE_LENGTH: usize> Drop for AnySlice<T, MAX_SLICE_LENGTH> {
+    fn drop(&mut self) {
+        if self.slice_len > 0 {
```

Comment:
I'm curious to know what happens if we remove this `if`. It's a bit counter intuitive that we are still calling `alloc` when `slice_len = 0` but we don't call `dealloc`. I would expect some symmetry in those calls.

Reply:
Removing the `if` produces a failure of the form:
```
Check 35: std::alloc::dealloc.pointer_primitives.4
         - Status: FAILURE
         - Description: "pointer outside object bounds"
         - Location: ~/.rustup/toolchains/nightly-2022-03-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:105:14 in function std::alloc::dealloc
```
You're right about the asymmetry not looking correct: calling `alloc` with a zero size is actually undefined behavior (https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#safety-1). I'll see if I can fix that. Might need to store a null pointer in the case of a zero slice!

Reply:
Doesn't rust normally prefer to use a non-null but invalid value

Reply:
Indeed! `NonNull::dangling()` is the recommended approach on https://doc.rust-lang.org/core/slice/fn.from_raw_parts.html#safety.

Reply:
I added a commit that explicitly handles the `slice_len == 0` case via a `ptr::null()`.